### ISSUE TITLE
Is upgrade auto-finalization a good default?

### ISSUE THREAD

**Issue by nick-jones** [cockroachdb/cockroach#57887]
**Is your feature request related to a problem? Please describe.**

Twice now we've been in a position where moving to a new version has caused issues and a downgrade has been necessary:

- #48786
- #57032

In the first instance we unfortunately did not set `preserve_downgrade_option`. With the second issue we had set it, so managed to avoid any big catastrophe... though from the issue you can [someone else wasn't quite so lucky].(https://github.com/cockroachdb/cockroach/issues/57032#issuecomment-733122620)

The v20.2 upgrade documentation specifically states:

> we recommend disabling auto-finalization so you can monitor the stability and performance of the upgraded cluster before finalizing the upgrade

This suggests that _most people_ should be trying to avoid auto-finalization.

**Describe the solution you'd like**

My question is: is it sensible to default to auto-finalization, given some of the issues that pop up and the recommendations in your own documentation? I'm not actively watching issues here, so I don't have reasonable perspective on how often people get into a tangle as a result of this. I did, however, feel it was worth raising the question.

I fully understand requiring operators to take manual steps during upgrades is undesirable. I think it's worth weighing that up with fairly rapidly locking into a new version.

**Describe alternatives you've considered**

- Having a flexible downgrade path regardless of what has happened would be an option, though I suspect considerable effort
- Delay auto-finalization based on some default duration, perhaps considerable. This could be with some option to force finalization, when required.
- There is also the option to have a cluster-wide setting to disable auto-finalization permanently and allowing operators to manually finalize the upgrade (i.e. as I understand this was the old behaviour). If there is a general preferance to retain auto-finalization, then this can be defaulted to "off".

**Comment by ajwerner** [comment]
Thanks for raising the issue. It makes sense and is valid. One other note I'll make is that it's best to roll out the new version incrementally, by first testing one node and making sure everything looks good. Finalization cannot happen until all of the nodes are running the new version. 

Also, even if we don't change the default, I think we can do more to sanity check the upgrade. Namely, we should probably do more to validate the state of the cluster is stable. Perhaps by waiting to make sure nodes have been up for a little while and also to check that the schema seems valid.

For what it's worth, in the next release (21.1) we'll be introducing a long-running migration framework to perform durable migrations which today take multiple versions. These migrations will not be able to run until after the cluster version has been finalized. They may be yet another reason to revisit the user story here.

In short, I think making the default be to auto-finalize but after a significantly longer amount of time (days?) with the caveat that all nodes stay up during that period seems much more reasonable to me.

**Comment by ajwerner** [comment]
cc @vy-ton who's been thinking about the upgrade user story a bit lately.

**Comment by irfansharif** [comment]
Some internal discussion [here](https://cockroachlabs.slack.com/archives/C0HM2DZ34/p1607962425139900) and [here](https://cockroachlabs.slack.com/archives/C0HM2DZ34/p1607966555146700).
